graphql_compiler/schema_merging/rename_schema.py

```diff
@@ -0,0 +1,232 @@
+        raise SchemaError('Input schema does not define a valid schema.\n'
+                          'Message: {}'.format(e))
+
+    schema_data = get_schema_data(ast)
+    # Check that the input schema has no extensions
+    if schema_data.has_extension:
+        raise SchemaError('Input schema should not contain extensions.')
+
+    # Rename types, interfaces, enums
+    reverse_name_map = _rename_types(ast, rename_func, schema_data.query_type, schema_data.scalars)
+
+    # Rename root fields
+    reverse_root_field_map = _rename_root_fields(ast, rename_func, schema_data.query_type)
+
+    return RenamedSchema(schema_ast=ast, reverse_name_map=reverse_name_map,
+                         reverse_root_field_map=reverse_root_field_map)
+
+
+def _rename_types(ast, rename_func, query_type, scalars):
+    """Rename types, enums, interfaces using rename_func.
+
+    The query type will not be renamed. Scalar types, field names, enum values will not be renamed.
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename types, interfaces, enums, etc
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+        scalars: set of strings, the set of user defined scalars
+
+    Returns:
+        Dict[str, str], the new type name to original type name map
+
+    Raises:
+        SchemaError if the rename causes name conflicts
+    """
+    visitor = RenameSchemaVisitor(rename_func, query_type, scalars)
+    visit(ast, visitor)
+
+    return visitor.reverse_name_map
+
+
+def _rename_root_fields(ast, rename_func, query_type):
+    """Rename root fields, aka fields of the query type.
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename fields of the query type
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+
+    Returns:
+        Dict[str, str], the new root field name to original root field name map
+
+    Raises:
+        SchemaError if rename causes root field name conflicts
+    """
+    visitor = RenameRootFieldsVisitor(rename_func, query_type)
+    visit(ast, visitor)
+
+    return visitor.reverse_field_map
+
+
+class RenameSchemaVisitor(Visitor):
+    """Traverse a Document AST, editing the names of nodes."""
+    def __init__(self, rename_func, query_type, scalar_types):
+        self.rename_func = rename_func  # callable that takes string to string
+        self.reverse_name_map = {}  # Dict[str, str], from new name to original name
+        self.query_type = query_type  # str
+        self.scalar_types = frozenset(scalar_types)
+        self.builtin_types = frozenset({'String', 'Int', 'Float', 'Boolean', 'ID'})
+
+        # NOTE: behavior of Directive type is not completely clear
+        self.noop_types = frozenset({
```

Comment:
Please create a test, (not in this pr), where you assert that the three sets are disjoint.